File: services/feed/internal/domain/history.go
package domain

import "time"

const MaxHistory = 500

type HistoryEntry struct {
	UserID    string    `json:"user_id"`
	ArticleID string    `json:"article_id"`
	URL       string    `json:"url"`
	Title     string    `json:"title"`
	Source    string    `json:"source"`
	ViewedAt  time.Time `json:"viewed_at"`
	ViewCount int       `json:"view_count"`
}

func (h HistoryEntry) Normalized() (HistoryEntry, error) {
	snap, err := normalizeArticleSnapshot(h.UserID, h.ArticleID, h.URL, h.Title, h.Source)
	if err != nil {
		return HistoryEntry{}, err
	}
	h.UserID = snap.userID
	h.ArticleID = snap.articleID
	h.URL = snap.url
	h.Title = snap.title
	h.Source = snap.source
	if h.ViewCount < 1 {
		h.ViewCount = 1
	}
	return h, nil
}

File: apps/web/lib/history.ts
import { feedHeaders, feedURL } from "./feed-client";
import type { BookmarkInput } from "./bookmarks";

export type HistoryEntry = {
  user_id: string;
  article_id: string;
  url: string;
  title: string;
  source: string;
  viewed_at: string;
  view_count: number;
};

export async function listHistory(userID: string): Promise<HistoryEntry[]> {
  const res = await fetch(feedURL("/me/history"), {
    headers: feedHeaders(userID),
    cache: "no-store",
  });
  if (!res.ok) {
    throw new Error(`history list returned ${res.status}`);
  }
  const data = (await res.json()) as { history?: HistoryEntry[] };
  return data.history ?? [];
}

export async function recordHistory(userID: string, article: BookmarkInput): Promise<void> {
  const res = await fetch(feedURL("/me/history"), {
    method: "PUT",
    headers: feedHeaders(userID, { "Content-Type": "application/json" }),
    body: JSON.stringify(article),
    cache: "no-store",
  });
  if (!res.ok) {
    const detail = await res.text();
    throw new Error(detail.trim() || `history record returned ${res.status}`);
  }
}
